consumer/consumer.go

```diff
@@ -327,3 +365,87 @@ func storeMetadataReceivedEvent(pw *writer.ParquetWriter, event ethereum.Metadat
 		log.Trace().Msg("Wrote metadata_received event to Parquet file")
 	}
 }
+
+func (c *Consumer) HandleValidatorMetadataEvent() error {
+	for {
+		select {
+		case event := <-c.validatorMetadataChan:
+			c.log.Trace().Any("event", event).Msg("Received validator event")
+
+			maddr, err := ma.NewMultiaddr(event.Multiaddr)
+			if err != nil {
+				c.log.Error().Err(err).Msg("Invalid multiaddr")
+				continue
+			}
+
+			ip, err := maddr.ValueForProtocol(ma.P_IP4)
+			if err != nil {
+				ip, err = maddr.ValueForProtocol(ma.P_IP6)
+				if err != nil {
+					c.log.Error().Err(err).Msg("Invalid IP in multiaddr")
+					continue
+				}
+			}
+
+			portStr, err := maddr.ValueForProtocol(ma.P_TCP)
+			if err != nil {
+				c.log.Error().Err(err).Msg("Invalid port in multiaddr")
+				continue
+			}
+
+			port, err := strconv.Atoi(portStr)
+			if err != nil {
+				c.log.Error().Err(err).Msg("Invalid port number")
+				continue
+			}
+
+			isValidator := true
+			longLived := indexesFromBitfield(event.MetaData.Attnets)
+			shortLived := extractShortLivedSubnets(event.SubscribedSubnets, longLived)
+			// If there are no short lived subnets, then the peer is not a validator
+			if len(shortLived) == 0 {
+				isValidator = false
+			}
+
+			prevCache, found := c.validatorCache[event.ID]
+			prevNumObservations := uint64(0)
+			prevAvgValidatorCount := int32(0)
+			if found {
+				prevNumObservations = prevCache.NumObservations
+				prevAvgValidatorCount = prevCache.AverageValidatorCount
+			}
+
+			currValidatorCount := 1 + (len(shortLived)-1)/2
+			currAvgValidatorCount := ComputeNewAvg(prevAvgValidatorCount, prevNumObservations, currValidatorCount)
+			// If there are no short lived subnets, then the validator count is 0 and the average is the same as the previous
+			if len(shortLived) == 0 {
+				currValidatorCount = 0
+				currAvgValidatorCount = prevAvgValidatorCount
+			}
+
+			validatorMetadata := ch.ValidatorMetadataEvent{
+				PeerID:                event.ID,
+				ENR:                   event.ENR,
+				Multiaddr:             event.Multiaddr,
+				IP:                    ip,
+				Port:                  uint16(port),
+				LastSeen:              uint64(event.Timestamp),
+				LastEpoch:             uint64(event.Epoch),
+				PossibleValidator:     isValidator,
+				AverageValidatorCount: currAvgValidatorCount,
+				NumObservations:       prevNumObservations + 1,
+			}
+
+			c.validatorCache[event.ID] = &validatorMetadata
+
+			// Write to Clickhouse
+			if c.chClient != nil {
+				c.chClient.ValidatorEventChan <- &validatorMetadata
+				c.log.Trace().Any("validator_metadata", validatorMetadata).Msg("Inserted validator metadata")
+			}
+		default:
+			c.log.Debug().Msg("No validator metadata event")
+			time.Sleep(1 * time.Second) // Prevents busy waiting
```

Comment:
Read about this - busy waiting leads to more CPU utilization and energy consumption. (Not very useful here, can be in big scalable software)